File: iOS/ObjC/MonitorApp/MonitorApp/ConfigureViewController.m
//
//  ConfigureViewController.m
//  MonitorApp
//
//  Created by Radwar on 8/24/17.
//  Copyright © 2017 curbside. All rights reserved.
//

#import "ConfigureViewController.h"
#import "MonitorExtras.h"


@interface ConfigureViewController () <UITextFieldDelegate>

@property (strong, nonatomic) IBOutlet UITextField *siteIdentifierTextField;
@property (strong, nonatomic) IBOutlet UITextField *trackingIdentifierTextField;
@property (strong, nonatomic) IBOutlet UILabel *errorLabel;
@property (strong, nonatomic) IBOutlet UIButton *startUpdatesButton;

@end

@implementation ConfigureViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _startUpdatesButton.layer.cornerRadius = 4.0;
    _siteIdentifierTextField.delegate = self;
    _trackingIdentifierTextField.delegate = self;
    
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissKeyboard:)];
    [self.view addGestureRecognizer:tapGesture];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (IBAction)startUpdates:(id)sender {
    NSString *trackingIdentifier = _trackingIdentifierTextField.text;
    NSString *siteIdentifier = _siteIdentifierTextField.text;
    
    if (trackingIdentifier.length == 0) {
        _errorLabel.text = @"Please enter a valid tracking identifier.";
        return;
    }
    
    if (siteIdentifier.length == 0) {
        _errorLabel.text = @"Please enter a valid site identifier.";
        return;
    }
    
    // We are good. Set defaults and exit.
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setObject:trackingIdentifier forKey:kTrackingIdentifierKey];
    [userDefaults setObject:siteIdentifier forKey:kSiteIdentifierKey];
    [userDefaults synchronize];
    
    [self dismissViewControllerAnimated:YES completion:^{}];
}

- (void)dismissKeyboard:(UITapGestureRecognizer *)gesture
{
    [self.view endEditing:YES];
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if (textField == _trackingIdentifierTextField) {
        [_siteIdentifierTextField becomeFirstResponder];
    } else if (textField == _siteIdentifierTextField) {
        [self dismissKeyboard:nil];
        [self startUpdates:nil];
    } else {
        [self dismissKeyboard:nil];
    }
    return YES;
}

@end

File: iOS/MonitorApp/Swift/MonitorApp/ConfigureViewController.swift
//
//  ConfigureViewController.swift
//  MonitorApp
//
//  Created by Hon Chen on 9/7/17.
//  Copyright © 2017 curbside. All rights reserved.
//

import UIKit

class ConfigureViewController: UIViewController {
    
    @IBOutlet var siteIdentifierTextField: UITextField!
    @IBOutlet var trackingIdentifierTextField: UITextField!
    @IBOutlet var errorLabel: UILabel!
    @IBOutlet var startUpdatesButton: UIButton!

    override func viewDidLoad() {
        super.viewDidLoad()
        startUpdatesButton.layer.cornerRadius = 4.0
        siteIdentifierTextField.delegate = self
        trackingIdentifierTextField.delegate = self
        
        let tapGesture = UITapGestureRecognizer.init(target: self, action: #selector(dismissKeyboard))
        view.addGestureRecognizer(tapGesture)
    }
    
    @IBAction func startUpdates(_ sender: Any) {
        let trackingIdentifer = trackingIdentifierTextField.text ?? ""
        let siteIdentifier = siteIdentifierTextField.text ?? ""
        
        if trackingIdentifer.characters.count == 0 {
            errorLabel.text = "Please enter a valid tracking identifier."
            return
        }
        
        if siteIdentifier.characters.count == 0 {
            errorLabel.text = "Please enter a valid site identifier."
            return;
        }
        
        // We are good. Set defaults and exit.
        let userDefaults = UserDefaults.standard
        userDefaults.set(trackingIdentifer, forKey: kTrackingIdentifierKey)
        userDefaults.set(siteIdentifier, forKey: kSiteIdentifierKey)
        userDefaults.synchronize()
        
        dismiss(animated: true, completion: nil)
    }
    
    func dismissKeyboard() {
        view.endEditing(true)
    }

}

extension ConfigureViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if textField == trackingIdentifierTextField {
            siteIdentifierTextField.becomeFirstResponder()
        } else if textField == siteIdentifierTextField {
            dismissKeyboard()
            startUpdates(self)
        } else {
            dismissKeyboard()
        }
        return true
    }
}
